Pythonize epoch transitions (#711)
ethereum/eth2.0-specs: specs/core/0_beacon-chain.md

```python
{
    # Sender index
    'sender': 'uint64',
    # Recipient index

>     'to': 'uint64',

Reviewer: `to` -> `recipient`

```suggestion
    'recipient': 'uint64',
```

The steps below happen when `(state.slot + 1) % SLOTS_PER_EPOCH == 0`.

#### Helper functions
  
We define some helper functions:

```python
def get_current_attestations(state: BeaconState) -> List[PendingAttestation]:
    return  [a for a in state.latest_attestations if current_epoch == slot_to_epoch(a.data.slot)]
```

```python

> def get_current_total_balance(state: BeaconState):

Reviewer 2: ```suggestion
def get_current_total_balance(state: BeaconState) -> Gwei:
```

    return get_total_balance(state, get_active_validator_indices(state.validator_registry, current_epoch))
```

```python

> def get_previous_total_balance(state: BeaconState):

Reviewer 2: ```suggestion
def get_previous_total_balance(state: BeaconState) -> Gwei:
```

    return get_total_balance(state, get_active_validator_indices(state.validator_registry, previous_epoch))
```

```python
def get_previous_attestations(state: BeaconState) -> List[PendingAttestation]:

>     return  [a for a in state.latest_attestations if previous_epoch == slot_to_epoch(a.data.slot)]

Reviewer 2: Remove the redundant space
```suggestion
    return [a for a in state.latest_attestations if previous_epoch == slot_to_epoch(a.data.slot)]
```

```

```python

> def get_attesting_indices(attestations: List[PendingAttestation], state: BeaconState) -> List[int]:

Reviewer 2: Reorder the parameters and fix type hinting
```suggestion
def get_attesting_indices(state: BeaconState, attestations: List[PendingAttestation]) -> List[ValidatorIndex]:
```

    output = set({})
    for a in attestations:
        output = output.union([get_attestation_participants(state, a.data, a.aggregation_bitfield)])
    return sorted(list(output))
```

```python

> def get_attesting_balance(attestations: List[PendingAttestation], state: BeaconState) -> List[int]:

Reviewer 2: ```suggestion
def get_attesting_balance(state: BeaconState, attestations: List[PendingAttestation]) -> List[ValidatorIndex]:
```

> def get_attesting_balance(attestations: List[PendingAttestation], state: BeaconState) -> List[int]:

Reviewer 2: ditto, `SignedGwei`.

    return get_total_balance(state, get_attesting_indices(attestations, state))
```

```python

> def get_current_epoch_boundary_attestations(state: BeaconState):

Reviewer 2: ```suggestion
def get_current_epoch_boundary_attestations(state: BeaconState) -> List[Attestation]:
```

    return [
        a for a in get_current_attestations(state) if 
        a.data.epoch_boundary_root == get_block_root(state, get_epoch_start_slot(get_current_epoch(state)))
    ]
```

```python

> def get_previous_epoch_boundary_attestations(state: BeaconState):

Reviewer 2: ```suggestion
def get_previous_epoch_boundary_attestations(state: BeaconState) -> List[Attestation]:
```

    return [
        a for a in get_previous_attestations(state) if 
        a.data.epoch_boundary_root == get_block_root(state, get_epoch_start_slot(get_previous_epoch(state)))
    ]
```

**Note**: Total balances computed for the previous epoch might be marginally different than the actual total balances during the previous epoch transition. Due to the tight bound on validator churn each epoch and small per-epoch rewards/penalties, the potential balance difference is very low and only marginally affects consensus safety.

```python

> def get_winning_root_and_participants(state: BeaconState, shard: Shard):

Reviewer 2: (Oh I knew it should return the root and the participants!!!)

```suggestion
def get_winning_root_and_participants(state: BeaconState, shard: Shard) -> Tuple[Bytes32, List[ValidatorIndex]]:
```

    all_attestations = get_current_attestations(state) + get_previous_attestations(state)
    valid_attestations = [
        a for a in all_attestations if a.data.latest_crosslink == state.latest_crosslinks[shard]
    ]
    all_roots = [a.data.crosslink_data_root for a in valid_attestations]
    
    def get_attestations_for(root):
        return [a for a in valid_attestations if a.data.crosslink_data_root == root]
        
    winning_root = max(all_roots, key=lambda r: get_attesting_balance(get_attestations_for(r)))
    
    return winning_root, get_attesting_indices(get_attestations_for(winning_root))
```

```python

> def earliest_attestation(state: BeaconState, validator_index: ValidatorIndex):

Reviewer 2: ```suggestion
def earliest_attestation(state: BeaconState, validator_index: ValidatorIndex) -> Attestation:
```

    return min([
        a for a in state.previous_epoch_attestations if
        validator_index in get_attestation_participants(state, a.data, a.aggregation_bitfield)
    ], key = lambda a: a.inclusion_slot)
```

```python

> def inclusion_slot(state: BeaconState, validator_index: ValidatorIndex):

Reviewer 2: ```suggestion
def inclusion_slot(state: BeaconState, validator_index: ValidatorIndex) -> Slot:
```

    return earliest_attestation(state, validator_index).inclusion_slot
```

```python

> def inclusion_distance(state: BeaconState, validator_index: ValidatorIndex):

Reviewer 2: ```suggestion
def inclusion_distance(state: BeaconState, validator_index: ValidatorIndex) -> int:
```

> def inclusion_distance(state: BeaconState, validator_index: ValidatorIndex):

Reviewer 2: No, the distance between slots is not `Slot` in the context.

> def inclusion_distance(state: BeaconState, validator_index: ValidatorIndex):

Reviewer: I agree.

Is `int` defined anywhere? (as AFAIK python defines this based on whether your OS is 32 or 64 bit)

> def inclusion_distance(state: BeaconState, validator_index: ValidatorIndex):

Reviewer 2: In here, `int` is the loose *any* int type that we use it for type hinting. We can also define `Uint64` and `Int64` for readability. (https://github.com/ethereum/eth2.0-specs/pull/695#issuecomment-468517960)

> def inclusion_distance(state: BeaconState, validator_index: ValidatorIndex):

Reviewer: I definitely think this is a good idea to be explicit for the sake standardisation across the languages. (Even if it is with catch-all `Uint64`s or whatever.

    attestation = earliest_attestation(state, validator_index)
    return attestation.inclusion_slot - attestation.data_slot
```

#### Justification

Run the following function:

```python

> def update_justification_and_finalization(state):

Reviewer 2: ```suggestion
def update_justification_and_finalization(state: BeaconState) -> None:
```

    if (bitfield >> 0) % 8 == 0b111 and state.justified_epoch == current_epoch - 2:
        state.finalized_epoch = state.justified_epoch
    # The 1st/2nd most recent epochs are both justified, the 1st using the 2nd as source        
    if (bitfield >> 0) % 4 == 0b11 and state.justified_epoch == current_epoch - 1:
        state.finalized_epoch = state.justified_epoch
        
    # Rotate justified epochs
    state.previous_justified_epoch = state.justified_epoch
    state.justified_epoch = new_justified_epoch
```

#### Crosslinks

Run the following function:

```python
for slot in range(get_epoch_start_slot(previous_epoch), get_epoch_start_slot(next_epoch)):
    for crosslink_committee, shard in get_crosslink_committees_at_slot(state, slot):
        winning_root, participants = get_winning_root_and_participants(state, shard)
        if 3 * total_balance(participants) >= 2 * total_balance(crosslink_committee):
            state.latest_crosslinks[shard] = Crosslink(
                epoch=slot_to_epoch(slot),
                crosslink_data_root=winning_root
            )
```

#### Eth1 data

Run the following function:

```python

> def maybe_reset_eth1_period(state):

Reviewer 2: ```suggestion
def maybe_reset_eth1_period(state: BeaconState) -> None:
```

    if (get_current_epoch(state) + 1) % EPOCHS_PER_ETH1_VOTING_PERIOD == 0:
        for eth1_data_vote in state.eth1_data_votes:
            # If a majority of all votes were for a particular eth1_data value,
            # then set that as the new canonical value
            if eth1_data_vote.vote_count * 2 > EPOCHS_PER_ETH1_VOTING_PERIOD * SLOTS_PER_EPOCH:
                state.latest_eth1_data = eth1_data_vote.eth1_data
        state.eth1_data_votes = []
```

#### Rewards and penalties

First, we define some additional helpers:

```python

> def get_base_reward(state: BeaconState, index: ValidatorIndex) -> int:

Reviewer 2: ```suggestion
def get_base_reward(state: BeaconState, index: ValidatorIndex) -> Gwei:
```

    adjusted_quotient = integer_squareroot(get_previous_total_balance(state)) // BASE_REWARD_QUOTIENT
    return get_effective_balance(state, index) // adjusted_quotient // 5
```

```python

> def get_inactivity_penalty(state: BeaconState, index: ValidatorIndex) -> int:

Reviewer 2: ```suggestion
def get_inactivity_penalty(state: BeaconState, index: ValidatorIndex) -> Gwei:
```

    return (
        base_reward(state, index) +
        get_effective_balance(state, index) * epochs_since_finality // INACTIVITY_PENALTY_QUOTIENT // 2
    )
```

Note: When applying penalties in the following balance recalculations implementers should make sure the `uint64` does not underflow.

##### Justification and finalization

```python

> def get_justification_and_finalization_deltas(state: BeaconState) -> Dict[Int -> Int]:

Reviewer 2: ```suggestion
def get_justification_and_finalization_deltas(state: BeaconState) -> Dict[ValidatorIndex, int]:
```

>     previous_active_validator_indices = get_active_validator_indices(state.validator_registry, previous_epoch)

Reviewer 3: `previous_active_validator_indices` seems to be a duplicate of `active_validator_indices`

    epochs_since_finality = get_current_epoch(state) + 1 - state.finalized_epoch
    active_validator_indices = [
        i for i in range(len(state.validator_registry)) if
        is_active_validator(state.validator_registry[i], get_current_epoch(state) - 1)
    ]
    # Initialize deltas that we are returning to zero
    deltas = {index: 0 for index in range(len(state.validator_registry))}
    # Some helper variables
    boundary_attestations = get_previous_epoch_boundary_attestations(state)
    boundary_attesting_balance = get_attesting_balance(boundary_attestations)
    total_balance = get_previous_total_balance(state)
    total_attesting_balance = get_attesting_balance(get_previous_epoch_attestations(state))
    matching_head_attestations = [
        a for a in get_previous_epoch_attestations(state) if
        a.data.beacon_block_root == get_block_root(state, a.data.slot)
    ]
    matching_head_balance = get_attesting_balance(matching_head_attestations)
    # Normal case
    if epochs_since_finality <= 4:
        for index in active_validator_indices:
            # Expected FFG source

>             if index in get_attesting_indices(get_previous_epoch_attestations(state)):

Reviewer 2: Reminder: `get_previous_epoch_attestations` undefined. It was `previous_epoch_attestations = [a for a in state.latest_attestations if previous_epoch == slot_to_epoch(a.data.slot)]`.

                deltas[index] += get_base_reward(state, index) * total_attesting_balance // total_balance
                # Inclusion speed bonus
                deltas[index] += (
                    get_base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY //
                    inclusion_distance(state, index)
                )
            else:
                deltas[index] -= get_base_reward(state, index)
            # Expected FFG target
            if index in get_attesting_indices(boundary_attestations):
                deltas[index] += get_base_reward(state, index) * boundary_attesting_balance // total_balance
            else:
                deltas[index] -= get_base_reward(state, index)
            # Expected head

>             if index in matching_head_attestations:

Reviewer 2: credit to @NIC619:
```suggestion
            if index in get_attesting_indices(matching_head_attestations):
```

>             if index in matching_head_attestations:

Reviewer 3: ```suggestion
            if index in get_attesting_indices(matching_head_attestations):
```

            else:
                deltas[index] -= get_base_reward(state, index)    
            # Proposer bonus
            proposer_index = get_beacon_proposer_index(state, inclusion_slot(state, index))
            deltas[proposer_index] += base_reward(state, index) // ATTESTATION_INCLUSION_REWARD_QUOTIENT
    # Quadratic leak case
    else:
        for index in active_validator_indices:
            if index not in get_attesting_indices(get_previous_epoch_attestations(state)):
                deltas[index] -= get_inactivity_penalty(state, index, epochs_since_finality)
            else:
                # If a validator did attest, apply a small penalty for getting attestations included late
                deltas[index] += (
                    base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY //
                    inclusion_distance(state, index)
                )
                deltas[index] -= base_reward(state, index)
                    
            if index not in get_attesting_indices(boundary_attestations):
                deltas[index] -= get_inactivity_penalty(state, index, epochs_since_finality)

>             if index not in matching_head_attestations:

Reviewer 3: ```suggestion
            if index not in get_attesting_indices(matching_head_attestations):
```

                deltas[index] -= get_base_reward(state, index)
        for index in range(len(state.validator_registry)):

>             if index not in active_validator_indices and state.validator_registry[index].slashed:

Reviewer 3: ```suggestion
            if index in active_validator_indices and state.validator_registry[index].slashed:
```

Author: That clause is specifically there for inactive validators; active validators get slashed in the section right above.

Reviewer 3: Hmm… but the original condition states that `Any active validator index with validator.slashed is True, loses 2 * inactivity_penalty(state, index, epochs_since_finality) + base_reward(state, index).`.

Also the comment `# Penalize slashed-but-inactive validators as though they were active but offline` seems a little confusing to me or I probably have misunderstood what `active` means in this context.

Author: I actually think that's wrong and it should be any inactive and not-withdrawn validator loses that amount. The intention is to ensure that you can't get away from inactivity penalties by self-slashing to instantly log yourself out; you still have to endure them.

                deltas[index] -= (
                    2 * get_inactivity_penalty(state, index, epochs_since_finality) +
                    get_base_reward(state, index)
                )
    return deltas
```

##### Crosslinks

```python

> def get_crosslink_reward_deltas(state: BeaconState) -> Dict[int, int]:

Reviewer 2: ```suggestion
def get_crosslink_reward_deltas(state: BeaconState) -> Dict[Gwei, Gwei]:
```

> def get_crosslink_reward_deltas(state: BeaconState) -> Dict[int, int]:

Reviewer 2: Ah, I made a similar comment for `get_justification_and_finalization_deltas`... but, second through, it happened in Trinity too: the "deltas" might be negative number, which is not align with as the defination of that "`Gwei` is uint64". We ended up add another custom type `SignedGwei` in Trinity for this case.

    deltas = {index: 0 for index in range(len(state.validator_registry))}

>     for slot in range(get_epoch_start_slot(previous_epoch), get_epoch_start_slot(next_epoch)):

Reviewer 3: ```suggestion
    for slot in range(get_epoch_start_slot(previous_epoch), get_epoch_start_slot(current_epoch)):
```

        for crosslink_committee, shard in get_crosslink_committees_at_slot(state, slot):
            winning_root, participants = get_winning_root_and_participants(state, shard)
            participating_balance = get_total_balance(state, participants)
            total_balance = get_total_balance(state, committee)
            for index in crosslink_committee:
                if index in participants:
                    deltas[index] += get_base_reward(state, index) * participating_balance // total_balance
                else:
                    deltas[index] -= get_base_reward(state, index)
    return deltas
```

#### Apply rewards

Run the following:

```python

> def apply_rewards(state: BeaconState):

Reviewer 2: ```suggestion
def apply_rewards(state: BeaconState) -> None:
```

    deltas1 = get_justification_and_finalization_deltas(state)
    deltas2 = get_crosslink_deltas(state)
    for i in range(len(state.validator_registry)):

>         state.validator_balances[i] += deltas1[i] + deltas2[i]

Reviewer 3: Should it be `max`?

    ]
```

```python
def get_previous_epoch_matching_head_attestations(state: BeaconState) -> List[PendingAttestation]:
    return [
        a for a in get_previous_epoch_attestations(state) if
        a.data.beacon_block_root == get_block_root(state, a.data.slot)
    ]
```

**Note**: Total balances computed for the previous epoch might be marginally different than the actual total balances during the previous epoch transition. Due to the tight bound on validator churn each epoch and small per-epoch rewards/penalties, the potential balance difference is very low and only marginally affects consensus safety.

```python
def get_winning_root_and_participants(state: BeaconState, shard: Shard) -> Tuple[Bytes32, List[ValidatorIndex]]:
    all_attestations = get_current_attestations(state) + get_previous_attestations(state)
    valid_attestations = [
        a for a in all_attestations if a.data.latest_crosslink == state.latest_crosslinks[shard]
    ]
    all_roots = [a.data.crosslink_data_root for a in valid_attestations]
    
    def get_attestations_for(root) -> List[PendingAttestation]:
        return [a for a in valid_attestations if a.data.crosslink_data_root == root]
        

>     winning_root = max(all_roots, key=lambda r: get_attesting_balance(get_attestations_for(r)))

Reviewer 4: get_attesting_balance() missing 1 required positional argument: 'state'

    
    return winning_root, get_attesting_indices(get_attestations_for(winning_root))
```

```python
def earliest_attestation(state: BeaconState, validator_index: ValidatorIndex) -> PendingAttestation:
    return min([

>         a for a in state.previous_epoch_attestations if

Reviewer 2: ```suggestion
        a for a in state.latest_attestations if
```

Author: Maybe we should just move the logic from #697 into this PR? Would remove the need for the get_current/previous_epoch_attestations functions and clean a lot of this up, and if we're going to pythonize all the changes from #697 will need to be translated anyway.

Thoughts @JustinDrake @djrtwo?

        validator_index in get_attestation_participants(state, a.data, a.aggregation_bitfield)
    ], key=lambda a: a.inclusion_slot)
```

```python
def inclusion_slot(state: BeaconState, validator_index: ValidatorIndex) -> Slot:
    return earliest_attestation(state, validator_index).inclusion_slot
```

```python
def inclusion_distance(state: BeaconState, validator_index: ValidatorIndex) -> int:
    attestation = earliest_attestation(state, validator_index)
    return attestation.inclusion_slot - attestation.data.slot
```

#### Justification

Run the following function:

```python
def update_justification_and_finalization(state: BeaconState) -> None:

    new_justified_epoch = state.justified_epoch
    # Rotate the justification bitfield up one epoch to make room for the current epoch
    state.justification_bitfield <<= 1
    # If the previous epoch gets justified, fill the second last bit
    previous_boundary_attesting_balance = get_attesting_balance(get_previous_epoch_boundary_attestations(state))
    if previous_boundary_attesting_balance * 3 >= get_previous_total_balance(state) * 2:

>         new_justified_epoch = get_current_epoch(state) - 1

Reviewer 4: The `previous_epoch`s, which were defined by `get_previous_epoch`, are all replaced with `current_epoch -1` in this function body. This would be problematic in the genesis epoch.

Author: @djrtwo and I are talking about this; it seems likely that setting previous to equal current minus 1 in all cases could potentially be ok.

    # Crosslink data root is zero (to be removed in phase 1)
    assert attestation.data.crosslink_data_root == ZERO_HASH
    # Apply the attestation
    pending_attestation = PendingAttestation(
        data=attestation.data,
        aggregation_bitfield=attestation.aggregation_bitfield,
        custody_bitfield=attestation.custody_bitfield,

>         inclusion_slot=state.slot

Reviewer 2: nitpick
```suggestion
        inclusion_slot=state.slot,
```

    ]
```

```python
def get_previous_epoch_matching_head_attestations(state: BeaconState) -> List[PendingAttestation]:
    return [
        a for a in state.previous_epoch_attestations if
        a.data.beacon_block_root == get_block_root(state, a.data.slot)
    ]
```

**Note**: Total balances computed for the previous epoch might be marginally different than the actual total balances during the previous epoch transition. Due to the tight bound on validator churn each epoch and small per-epoch rewards/penalties, the potential balance difference is very low and only marginally affects consensus safety.

```python
def get_winning_root_and_participants(state: BeaconState, shard: Shard) -> Tuple[Bytes32, List[ValidatorIndex]]:
    all_attestations = state.current_epoch_attestations + state.previous_epoch_attestations
    valid_attestations = [
        a for a in all_attestations if a.data.latest_crosslink == state.latest_crosslinks[shard]
    ]
    all_roots = [a.data.crosslink_data_root for a in valid_attestations]
    
    def get_attestations_for(root) -> List[PendingAttestation]:
        return [a for a in valid_attestations if a.data.crosslink_data_root == root]
        

>     winning_root = max(all_roots, key=lambda r: get_attesting_balance(state, get_attestations_for(r)))

Reviewer 2: 1. It was noted that, in `get_winning_root`: ties broken by favoring lexicographically smallest `crosslink_data_root`.
2. In Python, [max() and min() on a list will return the first in the list that matches the max()/min() in the event of a tie](https://stackoverflow.com/a/6783051).
3. It seems now the tie break is depending on the order in `state.current_epoch_attestations + state.previous_epoch_attestations`? This change seems worth to be noted in release note.

/cc @NIC619

        else:
            # If a validator did attest, apply a small penalty for getting attestations included late
            deltas[index] += (
                base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY //
                inclusion_distance(state, index)
            )
            deltas[index] -= base_reward(state, index)
                    
        if index not in get_attesting_indices(boundary_attestations):
            deltas[index] -= get_inactivity_penalty(state, index, epochs_since_finality)
        if index not in get_attesting_indices(matching_head_attestations):
            deltas[index] -= get_base_reward(state, index)
    # Penalize slashed-but-inactive validators as though they were active but offline
    for index in range(len(state.validator_registry)):
        if index not in active_validator_indices and state.validator_registry[index].slashed:
            deltas[index] -= (
                2 * get_inactivity_penalty(state, index, epochs_since_finality) +
                get_base_reward(state, index)
            )
    return deltas
```

##### Crosslinks

```python

> def get_crosslink_deltas(state: BeaconState) -> List[Gwei]:

Reviewer 5: Agreed with @hwwhww to return a Tuple of lists of gwei (the positive and negative deltas in separate lists)

Reviewer 5: Done

    if (bitfield >> 1) % 8 == 0b111 and state.previous_justified_epoch == current_epoch - 3:
        state.finalized_epoch = state.previous_justified_epoch
    # The 2nd/3rd most recent epochs are both justified, the 2nd using the 3rd as source
    if (bitfield >> 1) % 4 == 0b11 and state.previous_justified_epoch == current_epoch - 2:
        state.finalized_epoch = state.previous_justified_epoch
    # The 1st/2nd/3rd most recent epochs are all justified, the 1st using the 3rd as source
    if (bitfield >> 0) % 8 == 0b111 and state.justified_epoch == current_epoch - 2:
        state.finalized_epoch = state.justified_epoch
    # The 1st/2nd most recent epochs are both justified, the 1st using the 2nd as source        
    if (bitfield >> 0) % 4 == 0b11 and state.justified_epoch == current_epoch - 1:
        state.finalized_epoch = state.justified_epoch
        
    # Rotate justified epochs
    state.previous_justified_epoch = state.justified_epoch
    state.justified_epoch = new_justified_epoch
```

#### Crosslinks

Run the following function:

```python
def process_crosslinks(state: BeaconState) -> None:
    current_epoch = get_current_epoch(state)

>     previous_epoch = current_epoch - 1

Reviewer 2: ```suggestion
    previous_epoch = get_previous_epoch(state)
```

    next_epoch = current_epoch + 1
    for slot in range(get_epoch_start_slot(previous_epoch), get_epoch_start_slot(next_epoch)):
        for crosslink_committee, shard in get_crosslink_committees_at_slot(state, slot):
            winning_root, participants = get_winning_root_and_participants(state, shard)
            participating_balance = get_total_balance(state, participants)
            total_balance = get_total_balance(state, crosslink_committee)
            if 3 * participating_balance >= 2 * total_balance:
                state.latest_crosslinks[shard] = Crosslink(
                    epoch=slot_to_epoch(slot),

>                     crosslink_data_root=winning_root

Reviewer 2: ```suggestion
                    crosslink_data_root=winning_root,
```

                )
```

#### Eth1 data

Run the following function:

```python
def maybe_reset_eth1_period(state: BeaconState) -> None:
    if (get_current_epoch(state) + 1) % EPOCHS_PER_ETH1_VOTING_PERIOD == 0:
        for eth1_data_vote in state.eth1_data_votes:
            # If a majority of all votes were for a particular eth1_data value,
            # then set that as the new canonical value
            if eth1_data_vote.vote_count * 2 > EPOCHS_PER_ETH1_VOTING_PERIOD * SLOTS_PER_EPOCH:
                state.latest_eth1_data = eth1_data_vote.eth1_data
        state.eth1_data_votes = []
```

#### Rewards and penalties

First, we define some additional helpers:

```python
def get_base_reward(state: BeaconState, index: ValidatorIndex) -> Gwei:
    if get_previous_total_balance(state) == 0:
        return 0

    adjusted_quotient = integer_squareroot(get_previous_total_balance(state)) // BASE_REWARD_QUOTIENT

>     return get_effective_balance(state, index) // adjusted_quotient // 5

Reviewer 6: do we want to do some constant folding upon the final parameter tuning for these numbers? same below w/ `INACTIVITY_PENALTY_QUOTIENT // 2` in `get_inactivity_penalty`

        domain=get_domain(state.fork, slot_to_epoch(transfer.slot), DOMAIN_TRANSFER)
    )
    # Process the transfer
    state.validator_balances[transfer.sender] -= transfer.amount + transfer.fee

>     state.validator_balances[transfer.to] += transfer.amount

Reviewer: `to` -> `recipient`
```suggestion
    state.validator_balances[transfer.recipient] += transfer.amount
```

    ]
```

```python
def get_previous_epoch_matching_head_attestations(state: BeaconState) -> List[PendingAttestation]:
    return [
        a for a in state.previous_epoch_attestations if
        a.data.beacon_block_root == get_block_root(state, a.data.slot)
    ]
```

**Note**: Total balances computed for the previous epoch might be marginally different than the actual total balances during the previous epoch transition. Due to the tight bound on validator churn each epoch and small per-epoch rewards/penalties, the potential balance difference is very low and only marginally affects consensus safety.

```python
def get_winning_root_and_participants(state: BeaconState, shard: Shard) -> Tuple[Bytes32, List[ValidatorIndex]]:
    all_attestations = state.current_epoch_attestations + state.previous_epoch_attestations
    valid_attestations = [
        a for a in all_attestations if a.data.latest_crosslink == state.latest_crosslinks[shard]
    ]
    all_roots = [a.data.crosslink_data_root for a in valid_attestations]

    # handle when no attestations for shard available
    if len(all_roots) == 0:
        return ZERO_HASH, []

>     def get_attestations_for(root) -> List[PendingAttestation]:

Reviewer 2: ```suggestion
    def get_attestations_for(root: Bytes32) -> List[PendingAttestation]:
```

        # Expected FFG source
        if index in get_attesting_indices(state, state.previous_epoch_attestations):
            deltas[0][index] += get_base_reward(state, index) * total_attesting_balance // total_balance
            # Inclusion speed bonus
            deltas[0][index] += (
                get_base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY //
                inclusion_distance(state, index)
            )
        else:
            deltas[1][index] += get_base_reward(state, index)
        # Expected FFG target
        if index in get_attesting_indices(state, boundary_attestations):
            deltas[0][index] += get_base_reward(state, index) * boundary_attesting_balance // total_balance
        else:
            deltas[1][index] += get_base_reward(state, index)
        # Expected head
        if index in get_attesting_indices(state, matching_head_attestations):
            deltas[0][index] += get_base_reward(state, index) * matching_head_balance // total_balance
        else:
            deltas[1][index] += get_base_reward(state, index)
        # Proposer bonus
        proposer_index = get_beacon_proposer_index(state, inclusion_slot(state, index))

>         deltas[0][proposer_index] += base_reward(state, index) // ATTESTATION_INCLUSION_REWARD_QUOTIENT

Reviewer: ```suggestion
        deltas[0][proposer_index] += get_base_reward(state, index) // ATTESTATION_INCLUSION_REWARD_QUOTIENT
```
base_reward() -> get_base_reward()

    return deltas
```

When blocks are not finalizing normally...
            
```python
def compute_inactivity_leak_deltas(state: BeaconState) -> Tuple[List[Gwei], List[Gwei]]:
    # deltas[0] for rewards
    # deltas[1] for penalties
    deltas = [
        [0 for index in range(len(state.validator_registry))],
        [0 for index in range(len(state.validator_registry))]
    ]
    boundary_attestations = get_previous_epoch_boundary_attestations(state)
    matching_head_attestations = get_previous_epoch_matching_head_attestations(state)
    active_validator_indices = get_active_validator_indices(state.validator_registry, get_previous_epoch(state))
    epochs_since_finality = get_current_epoch(state) + 1 - state.finalized_epoch
    for index in active_validator_indices:
        if index not in get_attesting_indices(state, state.previous_epoch_attestations):
            deltas[1][index] += get_inactivity_penalty(state, index, epochs_since_finality)
        else:
            # If a validator did attest, apply a small penalty for getting attestations included late
            deltas[0][index] += (

>                 base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY //

Reviewer: ```suggestion
                get_base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY //
```
base_reward() -> get_base_reward()

                inclusion_distance(state, index)
            )

>             deltas[1][index] += base_reward(state, index)

Reviewer: ```suggestion
            deltas[1][index] += get_base_reward(state, index)
```
base_reward() -> get_base_reward()
